src/lib/controllers/ui-controller.js

```diff
@@ -196,9 +198,20 @@ export default class UIController {
           }
           this.panelData.tabs[name] = true
           this.state.changeTab(name) // Reflect a tab change in a state
+
+          this.initGrammarTab()
```

Comment:
Do we need to call `initGrammarTab()` every time a tab is changed?
To my understanding, grammarTab needs to be updated only if:
1. A panel is open for the first time (i.e. grammar tab initialization). Handled on line 169.
2. A current language information is changed:
   - user selects a different preferred language. This should be handled by `updateLanguage()` within `settingChange()`.
   - when a lexical request is made.  In that case a lexical query will call `updateLanguage()` that should take care of requesting updated grammar resources.
Are there any scenarios that are not covered by the above?

I think we should keep number of resource queries to the minimum as those are expensive.

What do you think?

Reply:
It calls resourceQuery only if there are no lexemes and if grammarRes url is undefined - than it loads grammar res for preffered language - so there is no additional resource queries.
'''
initGrammarTab () {
          if (this.panelData.tabs['grammar'] &&
              (!this.panelData.grammarRes || (!this.panelData.grammarRes.url && this.panelData.lexemes.length === 0))) {
'''
On 169 line it inits grammar only in the case when it was opened on grammar tab as the first tab.
So if the first tab was info for example - it won't make resource query.
But if a user changes tab to grammar - and there was no ready grammar res or lexemes - it will provide a resource query once. 

In this workflow - initGrammarRes doesn't intersect with yours scenarios.

Reply:
I'm maybe not up to latest changes here, so I'm trying to understand a logic behind updating grammar on a tab change. Please correct me if I will make any mistakes.

I think we need to change grammar only if the current language changes, because grammar is related to the language.
So, I'm thinking, the logic should be as follows:
1. We need to request grammar when extension is activated for the first time. This should be handled by `updateLanguage()` call within `this.options.load()` callback.
2. We need to request grammar if user changes a preferred language. This should be handled by the `updateLanguage()` call within `settingChange()` callback.
3. We need to request grammar if there is a new word lookup. This is handled by the lexical query that calls `updateLanguage()`.
4. We need to request grammar every time grammar information is lost. This might happen in the case described in the issue, when user refreshes page thus erasing all panel state. However, in this case a panel will be closed, so an `updateLanguage()` called when panel is open should solve this (you've already added that to the code).

Are there any scenarios not covered by the above cases that require us to make `initGrammarRes()` on a tab change? What's the reason for use to track tab changes?

Your code is working properly by checking if we have any grammar resources available before making a resource query so it's safe here, but I'm wondering what could be the simplest, most straightforward solution we can use. I'm also thinking about the future refactoring :-).

The other thing is (this is probably an additional improvement, but probably it's worth making if we're touching this because it should be easy, @balmas, what do you think?): if grammar is related to the language, then we need to run a new grammar request only when language changes. So maybe we can store a current language, and run a query in `updateLanguage` only if language of a new lexical query does not match the current language and grammar needs to be updated? And if we would know that we run grammar request every time language changes, we can rely on language variable to see if grammar request is necessary: if language is undefined or does not match the one of the current request, we have no grammar data or it is outdated, so grammar needs to be requested. In all other cases we're fine. So, in simple terms, language === grammar. So we can check language to see if grammar is fine or not. What do you think?